### PR TITLE
Lt 1300 good ratings and descriptions on applications

### PR DESCRIPTION
**Pull request by Mark-Spiers-kainos**

### CODE REVIEW

File: cases/views/goods.py [hunk 1]
```diff
@@ -15,3 +19,76 @@ def get(self, request, **kwargs):
             'good': good['good']
         }
         return render(request, 'cases/case/good.html', context)
+
+
+class ReviewGoods(TemplateView):
+    def get(self, request, **kwargs):
+        case_id = str(kwargs['pk'])
+        objects = request.GET.getlist('items', request.GET.getlist('goods'))
```
Comment (BenAndrew-Kainos) [inline cases/views/goods.py]: Could you name this something other than `objects`

File: cases/views/goods.py [hunk 2]
```diff
@@ -15,3 +19,76 @@ def get(self, request, **kwargs):
...
+        form_data = {
+            'objects': self.goods,
+            'comment': request.POST.get('comment'),
+            'report_summary': request.POST.get('report_summary'),
+            'control_code': request.POST.get('control_code', None)
+        }
+
+        if request.POST.get('is_good_controlled') == 'yes':
+            form_data['is_good_controlled'] = True
+        elif request.POST.get('is_good_controlled') == 'no':
```
Comment (BenAndrew-Kainos) [inline cases/views/goods.py]: else would hopefully suffice

File: templates/cases/case/review-goods.html [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		{% endif %}
+	</a>
+</div>
+
+
+<!-- Case flags -->
+{#{% include 'cases/case/views/case-flags.html' %}#}
+
+<!-- Information board -->
+{#{% include 'cases/case/views/application-information-board.html' %}#}
```
Comment (BenAndrew-Kainos) [inline templates/cases/case/review-goods.html]: Remove this if no longer needed

File: ui_automation_tests/pages/good_summary_page.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        edit_goods_btn = self.driver.find_element_by_id(self.edit_goods_flags)
+        edit_goods_btn.click()
+
+    def click_add_report_summary(self):
+        return self.driver.find_element_by_id(self.add_report_summary).click()
+
+    def check_control_code(self):
+        rows = self.driver.find_elements_by_css_selector('.lite-table__body>.lite-table__row')
+        for row in rows:
+            assert 'ML4b1' in row.find_elements_by_css_selector('.lite-table__cell')[3].text
```
Comment (BenAndrew-Kainos) [inline ui_automation_tests/pages/good_summary_page.py]: Asserts shouldn't be in Pages. You should make this method return the rows text and then put the assert in the step def

File: ui_automation_tests/pages/good_summary_page.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+class GoodSummaryPage(BasePage):
+    edit_goods_flags = 'button-edit-goods-flags'  # ID
+    add_report_summary = 'button-add-report-summary' # ID
+
+    def click_edit_good_flags(self):
+        edit_goods_btn = self.driver.find_element_by_id(self.edit_goods_flags)
+        edit_goods_btn.click()
+
+    def click_add_report_summary(self):
+        return self.driver.find_element_by_id(self.add_report_summary).click()
```
Comment (BenAndrew-Kainos) [inline ui_automation_tests/pages/good_summary_page.py]: No point returning a .click()

File: cases/views/flags.py [hunk 6]
```diff
@@ -74,13 +90,14 @@ def get(self, request, **kwargs):
...
-                                                         'note': request.POST.get('note')
-                                                     })
+        data = {
+                'level': self.level,
+                'objects': self.objects,
+                'flags': request.POST.getlist('flags'),
+                'note': request.POST.get('note')
+                }
+
+        response, _ = put_flag_assignments(request, data)
```
[the remark this answers is not in the document]
Comment (Mark-Spiers-kainos) [reply cases/views/flags.py]: we just split up the data of this request, testing it manually it seems to work fine.

File: cases/views/flags.py [hunk 7]
```diff
@@ -55,11 +62,20 @@ def dispatch(self, request, *args, **kwargs):
...
+                self.url += '?goods=' + self.objects[0]
+        else:
+            if origin == 'review_goods':
+                origin = 'review good'
+                objects_url_suffix = "?"
 
-        flags = [Option(x['id'], x['name']) for x in flags]
-        self.url = reverse('cases:' + origin, kwargs=kwargs)
+                for pk in self.objects:
+                    objects_url_suffix += '&goods=' + pk
```
[the remark this answers is not in the document]
Comment (Mark-Spiers-kainos) [reply cases/views/flags.py]: It does, should be fixed now

File: cases/views/flags.py [hunk 8]
```diff
@@ -55,11 +62,20 @@ def dispatch(self, request, *args, **kwargs):
...
 
             # Origin is set to tell the form where to return to after submission or when back link is clicked
-            if origin == 'good':
-                kwargs = {'pk': case_id, 'good_pk': self.objects[0]}
+            if origin == 'review_goods':
+                origin = 'review good'
+                self.url += '?goods=' + self.objects[0]
+        else:
+            if origin == 'review_goods':
+                origin = 'review good'
```
[the remark this answers is not in the document]
Comment (Mark-Spiers-kainos) [reply cases/views/flags.py]: fix

File: cases/views/flags.py [hunk 9]
```diff
@@ -18,20 +18,26 @@ class AssignFlags(TemplateView):
...
             raise Http404
 
+        self.level = request.GET.get('level')
+
+        case_id = str(kwargs['pk'])
+        kwargs = {'pk': case_id}
+        origin = request.GET.get('origin', 'case')
+
+        if origin == 'good':
+            kwargs = {'pk': case_id, 'good_pk': self.objects[0]}
```
[the remark this answers is not in the document]
Comment (Mark-Spiers-kainos) [reply cases/views/flags.py]: fixed

File: cases/views/goods.py [hunk 10]
```diff
@@ -15,3 +20,82 @@ def get(self, request, **kwargs):
...
+            form_data['is_good_controlled'] = False
+
+        response = post(request, '/goods/controlcode/', form_data)
```
Comment (andytaylorkainos) [inline cases/views/goods.py]: can we put this into a service?

File: cases/services.py [hunk 11]
```diff
@@ -211,6 +212,10 @@ def get_goods_type(request, pk):
     data = get(request, GOODS_TYPE_URL + pk)
     return data.json(), data.status_code
 
+def post_goods_control_code(request, json):
+    #this function may return just a status code, so we do not split it into data/statuscode
```
Comment (andytaylorkainos) [inline cases/services.py]: should we not update the api to return both the status code with data?

File: ui_automation_tests/pages/good_summary_page.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    add_report_summary = 'button-add-report-summary' # ID
+
+    def click_edit_good_flags(self):
+        edit_goods_btn = self.driver.find_element_by_id(self.edit_goods_flags)
+        edit_goods_btn.click()
+
+    def click_add_report_summary(self):
+        self.driver.find_element_by_id(self.add_report_summary).click()
+
+    def get_table_rows(self):
```
Comment (phlptt) [inline ui_automation_tests/pages/good_summary_page.py]: i belive there is something in shared for this

File: ui_automation_tests/step_defs/test_review_goods.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    query_page.enter_a_comment(comment)
+    Shared(driver).click_submit()
+
+
+@then("the control list is present on goods review page")
+def check_control_list_code(driver):
+    good_summary_page = GoodSummaryPage(driver)
+    rows = good_summary_page.get_table_rows()
+    for row in rows:
+        assert 'ML4b1' in row.find_elements_by_css_selector('.lite-table__cell')[3].text
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_review_goods.py]: think its worth storing this as a variable in context in the previous step where its passed in as a variable

Comment (phlptt) [inline ui_automation_tests/step_defs/test_review_goods.py]: also, is it aalways going to be the third cell? whenever tests are set to things relative to the location on the page, they become flaky.

File: cases/views/flags.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply cases/views/flags.py]: ok

File: ui_automation_tests/pages/good_summary_page.py (hunk 12 above)
Comment (BenAndrew-Kainos) [reply ui_automation_tests/pages/good_summary_page.py]: ^^^

File: ui_automation_tests/step_defs/test_review_goods.py (hunk 13 above)
Comment (BenAndrew-Kainos) [inline ui_automation_tests/step_defs/test_review_goods.py]: As discussed move this into the page method as you should only need to do `assert 'ML4b1' in good_summary_page.get_table_rows()` here

**Pull request merged**